src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -337,18 +340,117 @@ public static int GetMaxDecodedFromUtf8Length(int length)
         /// hence can only be called once with all the data in the buffer.
         /// </returns>
-        public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int bytesWritten)
-        {
-            OperationStatus status = DecodeFromUtf8InPlaceCore(buffer, out bytesWritten, out uint sourceIndex);
-            Debug.Assert(status is OperationStatus.Done or OperationStatus.InvalidData, "These are the only statuses the method is coded to return.");
+        public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int bytesWritten) =>
+            DecodeFromUtf8InPlace(buffer, out bytesWritten, ignoreWhiteSpace: true);
 
-            if (status != OperationStatus.Done)
+        private static unsafe OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int bytesWritten, bool ignoreWhiteSpace)
+        {
+            if (buffer.IsEmpty)
             {
-                // The input may have whitespace, attempt to decode while ignoring whitespace.
-                status = DecodeWithWhiteSpaceFromUtf8InPlace(buffer, ref bytesWritten, (int)sourceIndex);
+                bytesWritten = 0;
+                return OperationStatus.Done;
             }
 
-            return status;
+            fixed (byte* bufferBytes = &MemoryMarshal.GetReference(buffer))
+            {
+                uint bufferLength = (uint)buffer.Length;
+                uint sourceIndex = 0;
+                uint destIndex = 0;
+
+                // only decode input if it is a multiple of 4
+                if (bufferLength != ((bufferLength >> 2) * 4))
+                {
+                    goto InvalidExit;
+                }
+                if (bufferLength == 0)
+                {
+                    goto DoneExit;
+                }
+
+                ref sbyte decodingMap = ref MemoryMarshal.GetReference(DecodingMap);
+
+                while (sourceIndex < bufferLength - 4)
+                {
+                    int result = Decode(bufferBytes + sourceIndex, ref decodingMap);
+                    if (result < 0)
+                    {
+                        goto InvalidExit;
+                    }
+
+                    WriteThreeLowOrderBytes(bufferBytes + destIndex, result);
+                    destIndex += 3;
+                    sourceIndex += 4;
+                }
+
+                uint t0 = bufferBytes[bufferLength - 4];
+                uint t1 = bufferBytes[bufferLength - 3];
+                uint t2 = bufferBytes[bufferLength - 2];
+                uint t3 = bufferBytes[bufferLength - 1];
+
+                int i0 = Unsafe.Add(ref decodingMap, (IntPtr)t0);
+                int i1 = Unsafe.Add(ref decodingMap, (IntPtr)t1);
+
+                i0 <<= 18;
+                i1 <<= 12;
+
+                i0 |= i1;
+
+                if (t3 != EncodingPad)
+                {
+                    int i2 = Unsafe.Add(ref decodingMap, (IntPtr)t2);
+                    int i3 = Unsafe.Add(ref decodingMap, (IntPtr)t3);
```

Comment:
Is the cast to `IntPtr` necessary when `t{2,3}` is of type `uint` (so that Unsafe.Add will use the `nuint`-overload implicitly)?
```suggestion
                    int i2 = Unsafe.Add(ref decodingMap, t2);
                    int i3 = Unsafe.Add(ref decodingMap, t3);
```